src/sources/audiosourcemp3.cpp

```diff
@@ -0,0 +1,445 @@
+
+inline AudioSource::sample_type madScale(mad_fixed_t sample) {
+    return sample * kMadScale;
+}
+
+// Optimization: Reserve initial capacity for seek frame list
+const AudioSource::size_type kMinutesPerFile = 10; // enough for the majority of files (tunable)
+const AudioSource::size_type kSecondsPerMinute = 60; // fixed
+const AudioSource::size_type kMaxMp3FramesPerSecond = 39; // fixed: 1 MP3 frame = 26 ms -> ~ 1000 / 26
+const AudioSource::size_type kSeekFrameListCapacity = kMinutesPerFile
+        * kSecondsPerMinute * kMaxMp3FramesPerSecond;
+
+bool mad_skip_id3_tag(mad_stream* pStream) {
+    long tagsize = id3_tag_query(pStream->this_frame,
+            pStream->bufend - pStream->this_frame);
+    if (0 < tagsize) {
+        mad_stream_skip(pStream, tagsize);
+        return true;
+    } else {
+        return false;
+    }
+}
+
+} // anonymous namespace
+
+AudioSourceMp3::AudioSourceMp3(QString fileName) :
+        m_file(fileName),
+        m_fileSize(0),
+        m_pFileData(NULL),
+        m_avgSeekFrameCount(0),
+        m_curFrameIndex(0),
+        m_madSynthCount(0) {
+    mad_stream_init(&m_madStream);
+    mad_stream_options(&m_madStream, MAD_OPTION_IGNORECRC);
+    mad_frame_init(&m_madFrame);
+    mad_synth_init(&m_madSynth);
+    m_seekFrameList.reserve(kSeekFrameListCapacity);
+}
+
+AudioSourceMp3::~AudioSourceMp3() {
+    close();
+}
+
+AudioSourcePointer AudioSourceMp3::create(QString fileName) {
+    QSharedPointer<AudioSourceMp3> pAudioSource(new AudioSourceMp3(fileName));
+    if (OK == pAudioSource->open()) {
+        // success
+        return pAudioSource;
+    } else {
+        // failure
+        return AudioSourcePointer();
+    }
+}
+
+void AudioSourceMp3::addSeekFrame(
+        mad_timer_t madDuration,
+        mad_units madUnits,
+        const unsigned char* pInputData) {
+    SeekFrameType seekFrame;
+    seekFrame.pInputData = pInputData;
+    seekFrame.frameIndex = mad_timer_count(madDuration, madUnits);
+    m_seekFrameList.push_back(seekFrame);
+}
+
+Result AudioSourceMp3::open() {
+    if (!m_file.open(QIODevice::ReadOnly)) {
+        qWarning() << "Failed to open file:" << m_file.fileName();
+        return ERR;
+    }
+
+    // Get a pointer to the file using memory mapped IO
+    m_fileSize = m_file.size();
+    m_pFileData = m_file.map(0, m_fileSize);
+    // Transfer it to the mad stream-buffer:
+    mad_stream_buffer(&m_madStream, m_pFileData, m_fileSize);
+    DEBUG_ASSERT(m_pFileData == m_madStream.this_frame);
+
+    // Decode all the headers and calculate audio properties
+    unsigned long sumBitrate = 0;
+    setChannelCount(kChannelCountDefault);
+    setFrameRate(kFrameRateDefault);
+
+    mad_units madUnits = MAD_UNITS_44100_HZ; // default value
+    mad_timer_t madDuration = mad_timer_zero;
+
+    // Add first frame to list of frames
+    addSeekFrame(madDuration, madUnits, m_madStream.this_frame);
+
+    DEBUG_ASSERT(quint64(m_madStream.bufend - m_madStream.this_frame) == m_fileSize);
```

Comment:
Both m_madStream.bufend and m_madStream.this_frame must be initialized in mad_stream_buffer(). Otherwise your and the existing code are incorrect, too ;) Both members are accessed in the while-loop before the first invocation of mad_header_decode() in SoundSourceMp3::open().